python/dgl/nn/pytorch/conv/cugraph_relgraphconv.py

```diff
@@ -0,0 +1,388 @@
+"""Torch Module for Relational graph convolution layer using the
+aggregation primitives in cugraph-ops"""
+# pylint: disable=no-member, arguments-differ, invalid-name, too-many-arguments
+import math
+
+import torch as th
+from torch import nn
+
+try:
+    from pylibcugraphops.legacy.aggregators.node_level import (
+        agg_hg_basis_post_bwd_int32,
+        agg_hg_basis_post_bwd_int64,
+        agg_hg_basis_post_fwd_int32,
+        agg_hg_basis_post_fwd_int64,
+    )
+    from pylibcugraphops.legacy.structure.graph_types import (
+        message_flow_graph_hg_csr_int32,
+        message_flow_graph_hg_csr_int64,
+    )
+except ModuleNotFoundError:
+    raise ModuleNotFoundError(
+        "dgl.nn.CuGraphRelGraphConv requires pylibcugraphops to be installed."
+    )
+
+CUDA_SM_PER_BLOCK = 49152
+
+
+class RelGraphConvAgg(th.autograd.Function):
+    r"""Custom autograd function for R-GCN aggregation layer that uses the
+    aggregation functions in cugraph-ops."""
+
+    @staticmethod
+    def forward(ctx, g, fanout, num_rels, edge_types, feat, coeff):
+        r"""Compute the forward pass of R-GCN aggregation layer.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph.
+        fanout : int
+            Maximum in-degree of nodes.
+        num_rels : int
+            Number of relations.
+        edge_types : torch.Tensor
+            A 1D tensor of edge types. Note that pylibcugraphops only accepts
+            edge_types in int32, so any input of other integer types will be casted
+            into int32, thus introducing some overhead. Pass in int32 tensor for
+            best performance.
+        feat : torch.Tensor, dtype=torch.float32, requires_grad=True
+            Tensor of source node features. Shape: (num_src_nodes, in_feat).
+        coeff : torch.Tensor, dtype=torch.float32, requires_grad=True
+            The coefficient matrix used in basis-decomposition regularization.
+            Shape: (num_rels, num_bases). It should be set to ``None``
+            when ``regularizer=None``.
```

Comment:
I am a little bit confused here. What will happen in regularizer is not "None" and coeff is also not "None"?

The CuGraphRelGraphConv calls RelGraphConvAgg, and regularizer is a property in CuGraphRelGraphConv. From the code structure POV, it is a little bit weird that a variable in a callee has a restriction depends on caller.

Reply:
I have updated the doc to avoid ambiguity here. `RelGraphConvAgg` does not use the `RelGraphConvAgg` variable at all, as the regularization information is carried by coeff itself.